### PR TITLE
Use native Ginkgo test runner instead of cmd/e2e (attempt N+1)

### LINKED ISSUE #8464: build sometimes fails with "mv: cannot stat 'e2e.test': No such file or directory"
My recent PR #8428 builds the e2e test as a binary and moves it into the output directory (https://github.com/GoogleCloudPlatform/kubernetes/blob/master/hack/lib/golang.sh#L367-371).

This sometimes works without issue, and other times fails with this error message:
```
mv: cannot stat 'e2e.test': No such file or directory
!!! Error in /go/src/github.com/GoogleCloudPlatform/kubernetes/hack/lib/golang.sh:371
  'mv -f "$(basename ${outfile})" "${outfile}"' exited with status 1
Call stack:
  1: /go/src/github.com/GoogleCloudPlatform/kubernetes/hack/lib/golang.sh:371 kube::golang::build_binaries_for_platform(...)
  2: /go/src/github.com/GoogleCloudPlatform/kubernetes/hack/lib/golang.sh:468 kube::golang::build_binaries(...)
  3: hack/build-cross.sh:31 main(...)
Exiting with status 1
```

I have no idea why this is flaky, but hoping that by opening an issue others can help debug (or mention if they have a more consistent repro case).

### PR DESCRIPTION
**Pull request by ixdy**
Yet another attempt to get this merged. This time I've addressed the issues in #8464.

Since `go test -c` saves the binary to the current working directory, this can cause failures when multiple builds are run in parallel (such as by `build/release.sh`). Using a lockfile to serialize calls to `go test -c` should fix this issue.

### CODE REVIEW

**Comment by vmarmol** [comment]
You'll need to rebase btw

**Comment by ixdy** [comment]
OK, rebased and reverted the revert once again - this should be ready for review.

The changes since the last attempt to merge the Ginkgo runner are viewable in https://github.com/ixdy/kubernetes/commit/5a4a756871d39a3ed1d318a3a7fca79069856171.

**Comment by ixdy** [comment]
cc @quinton-hoole

**Comment by ixdy** [comment]
As discussed on chat, this now has a much simpler fix - cd into the output directory before running `go test -c`. See the diff in https://github.com/ixdy/kubernetes/commit/71b9316a2f2638602c342982eee92748a1d606c6.

**Comment by quinton-hoole** [comment]
Just for completeness, here's a transcripts of the chat session...

@quinton-hoole: "Perhaps it's possible to change the working directory per invocation of "test-c".  Just a thought.  Possibly a lot simpler than all of the file locking stuff."

@quinton-hoole: "Also, assuming that you do in fact need the lock files, any reason you didn't use the more standard approach: http://linux.die.net/man/1/flock ?"

@ixdy: "OS X doesn't have flock"

@quinton-hoole: "On OS X you can use shlock? https://developer.apple.com/library/mac/documentation/Darwin/Reference/ManPages/man1/shlock.1.html
"

@ixdy: "I tried shlock, but that wasn't in our docker image. I wanted to avoid having both flock and shlock"

File: hack/lib/golang.sh [hunk 1]
```diff
@@ -349,6 +358,20 @@ kube::golang::build_binaries_for_platform() {
     fi
   fi
 
+  for test in "${tests[@]:+${tests[@]}}"; do
+    local outfile=$(kube::golang::output_filename_for_binary "${test}" \
+      "${platform}")
+    # Go 1.4 added -o to control where the binary is saved, but Go 1.3 doesn't
+    # have this flag. Whenever we deprecate go 1.3, update to use -o instead of
+    # changing into the output directory.
+    local orig_path=$(pwd)
```
Comment (quinton-hoole) [inline hack/lib/golang.sh]: nit: Neater to use pushd and popd, which does this for you.  Seems to work on linux, mac os x, windows.  But I guess that introduces an extra dependency. Up to you.

**Comment by quinton-hoole** [comment]
One nit for you to consider.  LGTM!

File: hack/lib/golang.sh (hunk 1 above)
Comment (ixdy) [inline hack/lib/golang.sh]: I like it. done.

**Comment by zmerlynn** [comment]
LGTM

**Comment by quinton-hoole** [comment]
Can be merged on Travis or Shippable green.

**Comment by ixdy** [comment]
Shippable is green.

**Comment by mbforbes** [comment]
Well, everyone seems happy with this to be merged, and I'm waiting on it to do work, so I'm just going to merge it.

**Comment by mbforbes** [comment]
Jenkins has been somewhat unhappy since this went in, and I'm seeing panics on my e2es locally—is this related?

```
Services
  should be able to create a functioning external load balancer
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:341
[BeforeEach] Services
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:53
>>> testContext.KubeConfig: /Users/maxforbes/.kube/config
STEP: Building a namespace api objects
[It] should be able to create a functioning external load balancer
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:341
STEP: creating service external-lb-test with external load balancer in namespace e2e-tests-service-0-b2ceae55-1952-479e-a1c2-da1ecbcdb8de
STEP: waiting up to 4m0s for service external-lb-test in namespace e2e-tests-service-0-b2ceae55-1952-479e-a1c2-da1ecbcdb8de to have a public IP
INFO: Waiting for service external-lb-test in namespace e2e-tests-service-0-b2ceae55-1952-479e-a1c2-da1ecbcdb8de to have a public IP (5.043494798s)
INFO: Waiting for service external-lb-test in namespace e2e-tests-service-0-b2ceae55-1952-479e-a1c2-da1ecbcdb8de to have a public IP (10.043818265s)
INFO: Waiting for service external-lb-test in namespace e2e-tests-service-0-b2ceae55-1952-479e-a1c2-da1ecbcdb8de to have a public IP (15.051366025s)
INFO: Waiting for service external-lb-test in namespace e2e-tests-service-0-b2ceae55-1952-479e-a1c2-da1ecbcdb8de to have a public IP (20.043517621s)
INFO: Waiting for service external-lb-test in namespace e2e-tests-service-0-b2ceae55-1952-479e-a1c2-da1ecbcdb8de to have a public IP (25.043252354s)
INFO: Waiting for service external-lb-test in namespace e2e-tests-service-0-b2ceae55-1952-479e-a1c2-da1ecbcdb8de to have a public IP (30.044117793s)
INFO: Waiting for service external-lb-test in namespace e2e-tests-service-0-b2ceae55-1952-479e-a1c2-da1ecbcdb8de to have a public IP (35.043746374s)
STEP: deleting service external-lb-test in namespace e2e-tests-service-0-b2ceae55-1952-479e-a1c2-da1ecbcdb8de
[AfterEach] Services
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:62
STEP: Destroying namespace e2e-tests-service-0-b2ceae55-1952-479e-a1c2-da1ecbcdb8de
STEP: Destroying namespace e2e-tests-service-1-751cb92b-1682-49de-b4ed-f6f2f79c5645

•! Panic [40.735 seconds]
Services
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:390
  should be able to create a functioning external load balancer [It]
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:341

  Test Panicked
  runtime error: invalid memory address or nil pointer dereference
  /usr/src/go/src/runtime/panic.go:387

  Full Stack Trace
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:409 (0x76b3c)
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:283 (0xba111)
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/e2e_test.go:133 (0x8a3e3)
  /usr/src/go/src/testing/testing.go:447 (0x603df)
  /usr/src/go/src/runtime/asm_amd64.s:2232 (0x3ff01)
```

---

```
Services
  should correctly serve identically named services in different namespaces on different external IP addresses
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:389
[BeforeEach] Services
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:53
>>> testContext.KubeConfig: /Users/maxforbes/.kube/config
STEP: Building a namespace api objects
[It] should correctly serve identically named services in different namespaces on different external IP addresses
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:389
STEP: creating service s0 in namespace e2e-tests-service-0-34ff51b5-48ba-4ef2-8612-f73e2674973e
STEP: creating service s0 in namespace e2e-tests-service-1-7f1fc7a0-5c5f-4e8e-b1eb-b3e7014d52dc
STEP: waiting up to 4m0s for service s0 in namespace e2e-tests-service-0-34ff51b5-48ba-4ef2-8612-f73e2674973e to have a public IP
INFO: Waiting for service s0 in namespace e2e-tests-service-0-34ff51b5-48ba-4ef2-8612-f73e2674973e to have a public IP (5.043696358s)
INFO: Waiting for service s0 in namespace e2e-tests-service-0-34ff51b5-48ba-4ef2-8612-f73e2674973e to have a public IP (10.042700496s)
INFO: Waiting for service s0 in namespace e2e-tests-service-0-34ff51b5-48ba-4ef2-8612-f73e2674973e to have a public IP (15.045310131s)
INFO: Waiting for service s0 in namespace e2e-tests-service-0-34ff51b5-48ba-4ef2-8612-f73e2674973e to have a public IP (20.044477147s)
INFO: Waiting for service s0 in namespace e2e-tests-service-0-34ff51b5-48ba-4ef2-8612-f73e2674973e to have a public IP (25.044245035s)
INFO: Waiting for service s0 in namespace e2e-tests-service-0-34ff51b5-48ba-4ef2-8612-f73e2674973e to have a public IP (30.04345714s)
STEP: deleting service s0 in namespace e2e-tests-service-1-7f1fc7a0-5c5f-4e8e-b1eb-b3e7014d52dc
STEP: deleting service s0 in namespace e2e-tests-service-0-34ff51b5-48ba-4ef2-8612-f73e2674973e
[AfterEach] Services
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:62
STEP: Destroying namespace e2e-tests-service-0-34ff51b5-48ba-4ef2-8612-f73e2674973e
STEP: Destroying namespace e2e-tests-service-1-7f1fc7a0-5c5f-4e8e-b1eb-b3e7014d52dc

•! Panic [35.895 seconds]
Services
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:390
  should correctly serve identically named services in different namespaces on different external IP addresses [It]
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:389

  Test Panicked
  runtime error: invalid memory address or nil pointer dereference
  /usr/src/go/src/runtime/panic.go:387

  Full Stack Trace
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:409 (0x76b3c)
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:383 (0xbbd45)
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/e2e_test.go:133 (0x8a3e3)
  /usr/src/go/src/testing/testing.go:447 (0x603df)
  /usr/src/go/src/runtime/asm_amd64.s:2232 (0x3ff01)
```

Elasticsearch and kubectl tests are failing too, but perhaps those are known issues...

**Comment by ixdy** [comment]
Jenkins shows this panicking from before my change was merged. The elasticsearch failure is #8509 I think?

**Comment by quinton-hoole** [comment]
I just rolled back #8399 to resolve #8509.  It might well resolve the above panic problem too.  I've not looked too deeply into the cause of the panic though.

**Pull request merged**